File: 240623/마법의 숲 탐색/magical-forest-exploration.py
dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]


from collections import deque


def BFS(r, c):
    visited = [[False for _ in range(Y)] for _ in range(X)]

    result = r
    q = deque([(r, c)])
    visited[r][c] = True
    while q:
        x, y = q.popleft()

        for d in range(4):
            nx = x + dx[d]
            ny = y + dy[d]

            if nx < 0 or nx >= X or ny < 0 or ny >= Y: continue
            if board[nx][ny] == 0 or visited[nx][ny]: continue

            if board[nx][ny] == board[x][y] or is_exit[x][y]:
                q.append((nx, ny))
                visited[nx][ny] = True
                result = max(result, nx+1)

    return result


def reset():
    for x in range(X):
        for y in range(Y):
            board[x][y] = 0
            is_exit[x][y] = False


def move(i):
    global ans

    c = C[i]

    r = -2
    while r <= X-3:

        # 밑으로 내려갈 수 있음
        if board[r+2][c] == 0 and board[r+1][c-1] == 0 and board[r+1][c+1] == 0:
            r += 1

        # 서쪽 아래로
        elif (board[r+2][c] != 0 or board[r+1][c-1] != 0 or board[r+1][c+1] != 0) and c-2 >= 0 and r+2 <= X-1 and board[r-1][c-1] == 0 and board[r][c-2] == 0 and board[r+1][c-1] == 0 and board[r+1][c-2] == 0 and board[r+2][c-1] == 0:
            r += 1
            c -= 1
            D[i] = (D[i] + 3) % 4

        # 동쪽 아래로
        elif (board[r+2][c] != 0 or board[r+1][c-1] != 0 or board[r+1][c+1] != 0) and c+2 <= Y-1 and r+2 <= X-1 and board[r-1][c+1] == 0 and board[r][c+2] == 0 and board[r+1][c+1] == 0 and board[r+1][c+2] == 0 and board[r+2][c+1] == 0:
            r += 1
            c += 1
            D[i] = (D[i] + 1) % 4

        # 다 불가능
        else:
            if r-1 < 0:     # 골렘의 일부가 밖에
                reset()
                return
            break

    board[r][c] = i+1
    for j in range(4):
        board[r+dx[j]][c+dy[j]] = i+1
    is_exit[r+dx[D[i]]][c+dy[D[i]]] = True

    ans += BFS(r, c)


if __name__ == "__main__":
    X, Y, K = map(int, input().split())
    board = [[0 for _ in range(Y)]]
    for _ in range(X):
        board.append([0 for _ in range(Y)])
    C = []; D = []  # D: [북, 동, 남, 서]
    is_exit = [[False for _ in range(Y)] for _ in range(X)]
    ans = 0
    for _ in range(K):
        c, d = map(int, input().split())
        C.append(c-1)
        D.append(d)

    ans = 0
    for i in range(K):
        move(i)
    print(ans)

File: 240908/마법의 숲 탐색/magical-forest-exploration.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <cstring>

#define NORTH 0
#define EAST 1
#define SOUTH 2
#define WEST 3
#define MAX 75
#define MAX_K 1000
#define X first
#define Y second

using namespace std;

int R, C, K;
int grid[MAX][MAX];
int dr[4] = { -1, 0, 1, 0 };
int dc[4] = { 0, 1, 0, -1 };
int bot;
int dir[MAX_K];
bool isExit[MAX][MAX];

bool InRange(int r, int c) {
	return 1 <= r && r <= bot && 1 <= c && c <= C;
}

bool CanGoDown(int r, int c) {
	if (r >= bot - 1) return false;
	if (grid[r + 1][c - 1] != 0) return false;
	if (grid[r + 1][c] != 0) return false;
	if (grid[r + 1][c + 1] != 0) return false;
	if (grid[r + 2][c] != 0) return false;
	return true;
}

bool CanGoLeft(int r, int c) {
	if (!InRange(r, c - 2)) return false;
	if (!InRange(r + 2, c - 1)) return false;
	if (grid[r - 1][c - 1] != 0) return false;
	if (grid[r][c - 2] != 0) return false;
	if (grid[r + 1][c - 1] != 0) return false;
	if (grid[r + 1][c - 2] != 0) return false;
	if (grid[r + 2][c - 1] != 0) return false;
	return true;
}

bool CanGoRight(int r, int c) {
	if (!InRange(r, c + 2)) return false;
	if (!InRange(r + 2, c + 1)) return false;
	if (grid[r - 1][c + 1] != 0) return false;
	if (grid[r][c + 2] != 0) return false;
	if (grid[r + 1][c + 1] != 0) return false;
	if (grid[r + 1][c + 2] != 0) return false;
	if (grid[r + 2][c + 1] != 0) return false;
	return true;
}

bool TouchBot(int r) {
	return r + 1 == bot;
}

void GollemIsHere(int r, int c, int i) {
	grid[r][c] = i;
	grid[r - 1][c] = i;
	grid[r + 1][c] = i;
	grid[r][c - 1] = i;
	grid[r][c + 1] = i;
}

int bfs(int r, int c, int n) {
	queue<pair<int, int>> q;
	bool visited[MAX][MAX]; memset(visited, false, sizeof visited);
	vector<int> rows;

	q.push(make_pair(r, c));
	visited[r][c] = true;
	rows.push_back(r);
	while (!q.empty()) {
		auto cur = q.front(); q.pop();

		if (isExit[cur.X][cur.Y]) {
			for (int i = 0; i < 4; i++) {
				int nr = cur.X + dr[i];
				int nc = cur.Y + dc[i];

				if (!InRange(nr, nc)) continue;
				if (grid[nr][nc] == 0) continue;
				if (visited[nr][nc]) continue;

				q.push(make_pair(nr, nc));
				visited[nr][nc] = true;
				rows.push_back(nr);
			}
		}
		else {
			for (int i = 0; i < 4; i++) {
				int nr = cur.X + dr[i];
				int nc = cur.Y + dc[i];

				if (!InRange(nr, nc)) continue;
				if (grid[nr][nc] == 0) continue;
				if (grid[cur.X][cur.Y] != grid[nr][nc]) continue;
				if (visited[nr][nc]) continue;

				q.push(make_pair(nr, nc));
				visited[nr][nc] = true;
				rows.push_back(nr);
			}
		}
	}

	int ret = 0;
	for (int i = 0; i < (int)rows.size(); i++) {
		ret = max(ret, rows[i]);
	}
	return ret;
}

void reset() {
}

int Move(int r, int c, int n) {
	while (true) {
		if (CanGoDown(r, c)) {
			r++;
		}
		else if (CanGoLeft(r, c)) {
			r++;
			c--;
			dir[n] = (dir[n] - 1) % 4;
		}
		else if (CanGoRight(r, c)) {
			r++;
			c++;
			dir[n] = (dir[n] + 1) % 4;
		}
		else {
			if (r - 1 < 3) {
				memset(grid, 0, sizeof grid);
				memset(isExit, false, sizeof isExit);
				return 0;
			}
			GollemIsHere(r, c, n);
			isExit[r + dr[dir[n]]][c + dc[dir[n]]] = true;
			return bfs(r, c, n) - 3;
		}
	}
}

void print() {
	for (int r = 1; r <= bot; r++) {
		for (int c = 1; c <= C; c++) {
			cout << grid[r][c] << " ";
		}
		cout << endl;
	}
	cout << endl;
	for (int k = 0; k < K; k++) {
		cout << dir[k] << " ";
	}
	cout << endl;
	for (int r = 1; r <= bot; r++) {
		for (int c = 1; c <= C; c++) {
			cout << isExit[r][c] << " ";
		}
		cout << endl;
	}
}

int main() {
	cin >> R >> C >> K;
	bot = R + 3;

	int ans = 0;
	for (int gollem_num = 1; gollem_num <= K; gollem_num++) {
		int c, d;
		cin >> c >> d;
		dir[gollem_num] = d;

		int r = 2;
		ans += Move(r, c, gollem_num);

		// print();
	}
	cout << ans;

	return 0;
}